### ISSUE TITLE
Prefer `Vec::extend` over `Vec::append` if possible

### ISSUE THREAD

**Issue by oli-cosmian** [rust-lang/rust-clippy#4321]
If one collects and immediately `append`s, one should instead not collect and `extend` via the iterator instead.

```rust
    let mut v = vec![5, 6, 7];
    // any iterator.collect()
    let mut w: Vec<i32> = [1, 2, 3].iter().cloned().collect();
    v.append(&mut w);
```

should be

```rust
    let mut v = vec![5, 6, 7];
    // any iterator.collect()
    let w: Vec<i32> = [1, 2, 3].iter().cloned();
    v.extend(w);
```

**Comment by Compro-Prasad** [comment]
Any updates on this? Was looking for whether to use `Vec::extend` or `Vec::append` and landed here. Any reasons why I should prefer `Vec::extend`? The example you provided doesn't seem to explain what I was looking for.

**Comment by xFrednet** [comment]
Hey @Compro-Prasad, we now have a very similar lint called [`extend_with_drain`](https://rust-lang.github.io/rust-clippy/master/index.html#extend_with_drain) which basically checks for the same circumstances. 

The difference between the functions is, that `extend` adds items one after another while, `append` adds all of them at once. This allows the destination data structure to decide how to add them efficiently. However, `append` is a bit less flexible than `extend`, as you always add the entire second data structure while the iterator of `extend` can be used to filter items.

Therefore, I would say that this greatly depends on the use case. If you already have an iterator or need to perform any actions on the items before adding them, then it's usually better to use `extend`. But if you have the second data structure and only want to add it directly, then it's better to use `append`. :upside_down_face:

**Comment by Compro-Prasad** [comment]
I simply want to append efficiently the whole list. I hope `append` would do the job.
